### PR TITLE
Wrap d3 range functions to generate ticks in the reporting timezone

### LINKED ISSUE #8413: Mismatched timeseries axes and labels in a Bar chart
Using a custom question against a BigQuery view, I see a discrepancy between the timestamps reported in the _axes_ and _labels_ of a **Bar** chart.  Incidentally, a **Row** chart displays correctly against the same data shows matched labels correctly.

- Browser: Chrome v68.0.3440.106 (64-bit)
- OS: OS X (v10.13.4)
- Databases: Google BigQuery
- Metabase version: 0.30.0 (using Docker image `metabase/metabase:v0.30.0`)
- Hosted: on Google Kubernetes engine
- Internal database: hosted via Google Cloud SQL PostgreSQL

**Question definition**

 question_definition

**Bar Chart**
_Incorrectly showing events from 24th August against the 23rd axes label_

bar_mismatched_labels

**Row Chart**
_Correctly showing events from 24th August against the 24th axes label (apologies, the axes label is very feint)_
row_correct_labels

**Raw Data**

rawdata_correct_dates

**BigQuery - view data**
_Subset of data showing event_timestamp field stored as TIMESTAMP with timezone information_
bigquery_view_timestamp

Having read through some of the other issues referencing timezones as a potential cause (e.g. issues 2340, 5246 and PR 7247), I can confirm that the Metabase server is running in UTC timezone, Metabase "Report Timezone" is UTC, and the underlying raw data in Bigquery is stored with UTC timezone.  My browser timezone is currently set as BST (GMT+1) - manually setting this to an earlier timezone (e.g. Azores/GMT) fixes the display of the Bar chart (bars then appear under the correct axis label).

### PR DESCRIPTION
**Pull request by paulrosenzweig**
Resolves #8413

~The comment in the code contains most of the information. Essentially, we were calling `moment#utcOffset` and pushing tick timestamps into a different day before formatting them.~

We used `d3.time[unit]` to generate good locations for ticks. These ticks were on boundaries in local time, but we plotted the data in the reporting timezone. To format the ticks correctly we had to manipulate them.

The new approach taken in this PR is to instead produce ticks in the reporting timezone to begin with (by wrapping the d3 functions). That cleans up the formatting code and prevents issues with tick/data element alignment.

---

Potential issue: 
I'm converting to and from UTC. That might cause some issues when crossing daylight saving boundaries. I'm going to add tests for this before merging.

---

Another approach:
Would it instead make sense to do everything in UTC? (@salsakran floated this idea) The core thing we need is to have ticks match the data we're plotting. This PR tweaks the ticks, but it'd be a bit simpler to convert the data to UTC up front. Would that cause issues elsewhere?

### CODE REVIEW

File: frontend/src/metabase/visualizations/lib/apply_axis.js [hunk 1]
```diff
@@ -138,14 +139,31 @@ export function applyChartTimeseriesXAxis(
...
-        .format();
-      return formatValue(timestampFixed, {
-        ...chart.settings.column(dimensionColumn),
+    chart.xAxis().tickFormat(rawTimestamp => {
+      // `rawTimestamp` is a plain Date object which discards the timezone. If
+      // we're going to display a time, we need to add it back in so it's
+      // formatted correctly.
+      //
+      // An example:
+      // If we're passed "2019-08-15T03:00:00+03:00", but the data offset is -3
```
Comment (salsakran) [inline frontend/src/metabase/visualizations/lib/apply_axis.js]: perhaps a dumb question, but in what situations would the data offset and the timezone attached to a data point differ?

Is this only used if the Reporting Timezone and the User's browser timezone differ?

Comment (paulrosenzweig) [reply frontend/src/metabase/visualizations/lib/apply_axis.js]: We extract `dataOffset` from the first row returned form the server, so that's how we get reporting timezone. In this function though, d3 is passing us timestamps to format ticks in the user's local time.

> Is this only used if the Reporting Timezone and the User's browser timezone differ?

Yup! If those match, overwriting the offset has no effect.

**Comment by paulrosenzweig** [comment]
Handling daylight savings is more complicated that I thought. 🤦‍♂

> Nobody knew ~health care~ daylight savings could be so complicated.

# TL;DR

This code (and the previous code) assumes a fixed UTC offset for the data. This breaks when the data crosses a DST boundary. Really we should be assuming a fixed timezone. Users set a reporting timezone not a reporting UTC offset.

I think this should wait for a future PR. If people agree, I can create an issue (probably a P3?) from this comment.

# The Problem

Let's take this query as an example. The reporting timezone (and Metabase server) are set to "US/Eastern".

image

These are the rows returned from this query. Each timestamp has an explicit UTC offset, but we don't know the timezone.
```
"2019-03-08T00:00:00.000-05:00", 15                                                                                                                         
"2019-03-09T00:00:00.000-05:00", 14                                                                                                                         
"2019-03-10T00:00:00.000-05:00", 18                                                                                                                         
"2019-03-11T00:00:00.000-04:00", 10                                                                                                                         
"2019-03-12T00:00:00.000-04:00", 16                                                                                                                         
```

The rough procedure for generating ticks:
1. Take the bounds
(from "2019-03-08T00:00:00.000-05:00" to "2019-03-12T00:00:00.000-04:00")
2. Stretch the bounds a bit to provide margin
3. Select a sensible interval (days in this case)
4. Generate ticks for that range

Previously, we were generating those ticks in local time. With the code in this PR, we use d3's UTC range functions and then map the ticks back to the offset we saved (-05:00 here). That produces the following ticks:

```
2019-03-08T00:00:00.000-05:00
2019-03-09T00:00:00.000-05:00
2019-03-10T00:00:00.000-05:00
2019-03-11T00:00:00.000-05:00
2019-03-12T00:00:00.000-05:00
```

You can see this mismatch if you look closely at the 11th and 12th. The ticks are slightly off center from the bars.

image

# The Fix

I think we should mostly keep the procedure above, but with one change: map generated ticks back to a given timezone not an offset. 

d3 generated ticks:
```
2019-03-08T00:00:00.000Z
2019-03-09T00:00:00.000Z
2019-03-10T00:00:00.000Z
2019-03-11T00:00:00.000Z
2019-03-12T00:00:00.000Z
```

ticks mapped to an offset (-05:00):
```
2019-03-08T00:00:00.000-05:00
2019-03-09T00:00:00.000-05:00
2019-03-10T00:00:00.000-05:00
2019-03-11T00:00:00.000-05:00
2019-03-12T00:00:00.000-05:00
```

ticks mapped to an timezone (US/Eastern):
```
2019-03-08T00:00:00.000-05:00
2019-03-09T00:00:00.000-05:00
2019-03-10T00:00:00.000-05:00
2019-03-11T00:00:00.000-04:00
2019-03-12T00:00:00.000-04:00
```

Those ticks would line up with the data passed down from the server.

Note: While, the bars and ticks would align with this solution, they would be unevenly spaced. I think that's correct since in DST-affected timezones the days are themselves uneven.

## Solution Specifics

OK, but _how_ do we get the timezone rather than just the offset? I see three options:

1. Get it out of the store
We already have it stored in settings under `report-timezone`. 
2. Pass down the timezone with each query
It feels like we should couple the reporting timezone to the rows directly. The rows were produced for a specific reporting timezone. That timezone is relevant to their interpretation.
3. Determine timezone from row offsets
It's _probably_ technically possible, but this solution is beyond horrifying. I would chain myself to the codebase to prevent construction crews from implementing it.

I like (2) the most, but (1) seems like probably the easiest way to go.

# More Problems

For daily data, I think there is a correct solution, and we just need to implement it. It's more complicated for hourly (and more granular) data. The core problem: converting days to and from UTC is bijective, but converting hours is not.

If we have data in the range `2019-03-10T00:00:00-05:00` to `2019-03-10T03:00:00-04:00`, we'd generate ticks for three hours. However, only two hours have elapsed. 

Here I'm not sure what we can do besides recommend switching to a better reporting timezone. Thankfully it feels like an uncommon issue. Even if you have a reporting timezone with DST, you're probably not looking at granular data on the two nights a year when the offset changes.

**Comment by RMMSecurity** [comment]
> Handling daylight savings is more complicated that I thought. 🤦‍♂
> 
> > Nobody knew ~health care~ daylight savings could be so complicated.

This is actual gold.  😄

**Review by tlrobinson — COMMENTED** [review]
I also pointed this to you offline, but just so we have it on GitHub, #7247 was meant to handle timezones with DST. I think it basically does what you concluded the fix was "I think we should mostly keep the procedure above, but with one change: map generated ticks back to a given timezone not an offset" :)

**Comment by paulrosenzweig** [comment]
@tlrobinson yeah that's definitely the right way to do this. I should have read that PR more closely before I dove into this!

That said, if that PR will take a while more to merge, I think this fixes most of the problem and might be worth getting in. Currently, there are issues just with different browser timezones even without crossing DST boundaries. Otherwise, I'm happy to close in favor of the tz range solution.

**Comment by mazameli** [comment]
What's the status of this @paulrosenzweig @tlrobinson? Merging, needs changes, or closing in favor of a different solution?

**Comment by paulrosenzweig** [comment]
Closing in favor of a different solution. I'm not sure how much work it will take to get #7247 merged, but that's what we should od.
